### PR TITLE
changed @property (with ad-hoc hasattr caching) to instead use memoized-property library

### PR DESCRIPTION
**Pull request by iskandr**
Fixes https://github.com/hammerlab/pyensembl/issues/23

Also:
* check validity of gene_biotype on Gene object
* sped up `__contains__` and sequence access on `ReferenceTranscripts` by by-passing `pyfaidx.Fasta.__contains__` (which copies a list of keys every time) 

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/pyensembl/26)

### CODE REVIEW

File: pyensembl/gene.py [hunk 1]
```diff
@@ -43,40 +50,36 @@ def __str__(self):
...
-            self._transcripts = transcripts
-        return self._transcripts
+        # We're doing a SQL query for each transcript ID to fetch
+        # its particular information, might be more efficient if we
+        # just get all the columns here, but how do we keep that modular?
+        transcripts = [
+            Transcript(result[0], self.db)
+            for result in results
+        ]
+        self._transcripts = transcripts
```
Comment (ihodes) [inline pyensembl/gene.py]: I think you may be forgetting the `return self._transcripts`?

Comment (iskandr) [inline pyensembl/gene.py]: Good catch, this code actually works for arcane reasons (memoized_property will run this function to set self._transcripts and then return the value found there). Obviously need to change it to be sane.

File: pyensembl/transcript.py [hunk 2]
```diff
@@ -69,46 +69,40 @@ def __len__(self):
...
+        exons = [None] * len(results)
 
-                exon = Exon(exon_id, self.db)
-                exon_number = int(exon_number)
-                assert exon_number >= 1, "Invalid exon number: %s" % exon_number
-                assert exon_number <= len(exons), \
-                    "Invalid exon number: %s (max expected = %d)" % (
-                        exon_number, len(exons))
+        for entry in results:
+            exon_number, exon_id = entry
```
Comment (ihodes) [inline pyensembl/transcript.py]: Just FWIW you could do this destructuring in the loop, `for exon_number, exon_id in results:`

**Comment by ihodes** [comment]
LGTM % that one odd return statement

**Pull request merged**